### ISSUE TITLE
Compaction doesn't work as expected

### ISSUE THREAD

**Issue by andrewhathaway** [STRML/react-grid-layout#641]
Thanks for submitting an issue to RGL!

Please mark the type of this issue:

- [x] Bug
- [ ] Feature Request
- [ ] Question

I have a project currently using [this](https://www.npmjs.com/package/react-grid-layout-horizontal-compact) package due to the previous versions of RGL not support horizontal compaction. 

<img width="1117" alt="start" src="https://user-images.githubusercontent.com/1188830/31384166-a281e50e-adb5-11e7-8401-a69f951079ad.png">

When using this version, the compaction happens in way in which space is not wasted. Shown below:

<img width="760" alt="desired" src="https://user-images.githubusercontent.com/1188830/31384169-ab138a60-adb5-11e7-8b23-c1cbc5cb9cd4.png">

When using RGL `0.16.0`, the compaction is not as desired, and just lines up against one side. I've also found that is switches between left and right, randomly when state has been updated. Shown below:

<img width="784" alt="actual" src="https://user-images.githubusercontent.com/1188830/31384174-aeffbcde-adb5-11e7-8f03-753db8abe365.png">

I'd prefer it if compaction was just done like this, and blocks fit in to where they could?

Thanks.

**Comment by larrydahooster** [comment]
I can confirm it. The compacting seems to be broken after upgrade to 16.0.0 or it is a breaking change. I haven't played with

> compactType: ?('vertical' | 'horizontal') = 'vertical';
> preventCollision: ?boolean = false;

parameters yet, but the behaviour with default params seems broken in my opinion.

0.15.3 
![oct-16-2017 17-31-48](https://user-images.githubusercontent.com/4989851/31621621-7b576848-b29a-11e7-9998-731560c11029.gif)

0.16.0
![oct-16-2017 17-49-05](https://user-images.githubusercontent.com/4989851/31621694-a1bd23ba-b29a-11e7-8e2d-f15713810140.gif)

**Comment by larrydahooster** [comment]
I could also easily reproduce weird behaviour in the basic demo

![oct-16-2017 18-04-18](https://user-images.githubusercontent.com/4989851/31622407-9c7dde56-b29c-11e7-9a8a-a3d24a805991.gif)

**Comment by torkelo** [comment]
I am so seeing this. Feel strange that items flow up through full width panels

**Comment by andrewhathaway** [comment]
I just want it all to fit nicely together and work out where to position things, making it a nice grid of things, rather than a vertical line!

**Comment by torkelo** [comment]
So after debugging the behavior @larrydahooster  is seeing I found a number of issues. 

1) Compaction is not recursive and can easily cause items to flow through other items. Fixed this by making act of moving items back down after compaction recursive (and take into account collisions). Implementation here: https://github.com/grafana/react-grid-layout/blob/grafana/lib/utils.js#L181

2) The second problem was how moveElementAwayFromCollision found two issues here one I think introduced when horizontal compaction was introduced. 
This line https://github.com/STRML/react-grid-layout/blob/master/lib/utils.js#L327 causes fakeItem.y be collidesWith.y - ItemToMove.h which can be very low and cause itemToMove to jump over an intermediate item. If always set fakeItem.y to itemToMove.y 

I also had to change when moving element away from collision to take into account element hight and not just add +1 to y. 
https://github.com/grafana/react-grid-layout/blob/grafana/lib/utils.js#L365 

With all these changes I no longer see any strange behaviour of panels flowing or jumping through other panels. But unsure yet if I introduced other issues. Can open PR after more testing if the current behaviour (in v0.16.0) is not intended or desired.
